examples/n64lib/src/allocator.rs

```diff
@@ -0,0 +1,54 @@
+use crate::lock::interrupt_lock;
+use core::alloc::{GlobalAlloc, Layout};
+
+#[link(name = "c")]
+extern "C" {
+    fn malloc(size: usize) -> *mut u8;
+    fn free(ptr: *mut u8);
+}
+
+struct N64LibAlloc;
+
+unsafe impl GlobalAlloc for N64LibAlloc {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let _lock = interrupt_lock();
+
+        malloc(layout.size())
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        let _lock = interrupt_lock();
+
+        free(ptr);
+    }
+}
+
+#[global_allocator]
+static GLOBAL: N64LibAlloc = N64LibAlloc;
+
+#[alloc_error_handler]
+fn alloc_error(_layout: Layout) -> ! {
+    loop {}
```

Comment:
I learned a while back that this triggers undefined behavior in LLVM. 😭 https://github.com/rust-lang/rust/issues/28728

I don't know if there is a better thing to do than dropping into assembly to provably loop forever.

Reply:
That's a shame. 😒 Just added a utility method which performs the assembly hack, since it might be something we'll need to do other places too once the OS gets fleshed out.